### PR TITLE
`ci` automatic modlist.md

### PR DESCRIPTION
**Pull request by Quezler**
When one of the enigmatica repositories updates minecraftinstance.json it is basically impossible to know the change.

This draft will generate in the same format as `ModListCreator-3.0.0` in order to keep compatibility with the pwsh ones.

It will co-author the commit, with the person who last touched minecraftinstance.json alongside the bot:
Screen Shot 2022-06-15 at 23 19 06

The script is split into two parts so they can be used independently and don't strictly rely on being ran by ci:
Screen Shot 2022-06-15 at 23 19 15

At the time of writing this is just me stashing this draft overnight because it seems to work, cleanup is still planned.
Leaving this here now to gather reactions, perhaps niller doesn't even like the idea of the ci commiting to the repo.

Anyways i've noticed that during the modpack upload script this part takes an awful long time, so who knows... ^-^

(ps, this currently intentionally leaves the existing pwsh intact & doesn't mess with the changelog subdirectory)

### CODE REVIEW

**Review by NielsPilgaard — APPROVED** [review]
I like this a lot, including the commiting to the repo 😄 

I do have some requests though if you're willing to make some additions/changes:
- Have this workflow only work on pushes to the `develop` branch, and output to a `DEV-MODLIST.md` file instead.
- Make another very (*very*) similar workflow that only triggers on tags that match the regex `\d+.\d+.\d+`, which would be responsible for updating the `MODLIST.md`, and also create a modlist for the specific tag version in the changelogs folder.

What do you think? :)

File: automation/node/modlist-generate.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+console.log(readme[0]);
+
+const projects = [];
+
+function next(i) {
+    const installedAddon = minecraftinstance_json.installedAddons[i];
+    if (installedAddon) {
+
+        // this is the api behind the current modlist generator, not some custom one :)
+        const melanx_url = `https://curse.melanx.de/project/${installedAddon.addonID}`;
```
[the remark this answers is not in the document]
Comment (Quezler) [reply automation/node/modlist-generate.js:22]: https://github.com/ModdingX/ModListCreator/blob/024aa551030180ed7488e4c23614f753428ed294/src/main/java/org/moddingx/modlistcreator/ModListCreator.java#L26

File: automation/node/modlist-generate.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                //     distribution: true,
+                //     website: 'https://www.curseforge.com/minecraft/mc-mods/plonk',
+                //     thumbnail: 'https://media.forgecdn.net/avatars/230/201/637059414601035196.png',
+                //     file: { id: 3299406, name: 'plonk-1.16.5-9.0.8.jar' }
+                // }
+
+                next(i + 1); // process the next addon synchronously.
+            })
+            .catch(() => {
+                next(i); // the api has a tendency to be slow (or even time out), so in case of failure retry the current addon.
```
[the remark this answers is not in the document]
Comment (Quezler) [reply automation/node/modlist-generate.js:52]: If the api is slow/down for a long period of time, yes (the job has a timeout of 20 minutes tho)

**Comment by Quezler** [comment]
> Have this workflow only work on pushes to the develop branch, and output to a DEV-MODLIST.md file instead.

So then we would have two modlists, would it even make sense when it only triggers for the develop branch since that by definition is dev?

**Comment by Quezler** [comment]
> Make another very (very) similar workflow that only triggers on tags that match the regex \d+.\d+.\d+, which would be responsible for updating the MODLIST.md, and also create a modlist for the specific tag version in the changelogs folder.

Didn't give this part too much thought yet, kinda assumed you might still have wanted to do that part with your modpack uploader script (the copying of that file thats simply pulled from the develop branch from the root to the subdirectory) 🤔

File: automation/node/modlist-generate.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (NielsPilgaard) [reply automation/node/modlist-generate.js:52]: Ah right, got'cha.

**Comment by NielsPilgaard** [comment]
> > Have this workflow only work on pushes to the develop branch, and output to a DEV-MODLIST.md file instead.
> 
> So then we would have two modlists, would it even make sense when it only triggers for the develop branch since that by definition is dev?

Good point, no need to rename it.

**Comment by NielsPilgaard** [comment]
> > Make another very (very) similar workflow that only triggers on tags that match the regex \d+.\d+.\d+, which would be responsible for updating the MODLIST.md, and also create a modlist for the specific tag version in the changelogs folder.
> 
> Didn't give this part too much thought yet, kinda assumed you might still have wanted to do that part with your modpack uploader script (the copying of that file thats simply pulled from the develop branch from the root to the subdirectory) 🤔

I'd prefer it done with your action, if you're up for making it.. I'd prefer all the changelog / mod-changelog steps to be done through actions to be honest 😄

**Comment by Quezler** [comment]
> I'd prefer it done with your action, if you're up for making it.. I'd prefer all the changelog / mod-changelog steps to be done through actions to be honest 😄

I'll have to dig a little deeper into doing specific actions for specific tags, might automate it further in a future pull 🤔

**Comment by NielsPilgaard** [comment]
> > I'd prefer it done with your action, if you're up for making it.. I'd prefer all the changelog / mod-changelog steps to be done through actions to be honest 😄
> 
> I'll have to dig a little deeper into doing specific actions for specific tags, might automate it further in a future pull 🤔

I have an example :D
https://github.com/NielsPilgaard/Pilgaard.CronJobs/blob/master/.github/workflows/release.yml

**Review by NielsPilgaard — APPROVED** [review]
Alright, can I merge or would you like me to wait? :)

**Comment by Quezler** [comment]
mind if i give that a go in another pr? this one already has 51 commits and counting, and i prefer to work from clean diff slates 😅

**Comment by Quezler** [comment]
> Alright, can I merge or would you like me to wait? :)

merging now would be preferred ^-^

**Pull request merged**